### PR TITLE
State testing

### LINKED ISSUE #104: [Tracking issue] Extensive test set verifing if cheatable starknet behaviour is consisten with starknet behaviour
Things to consider: 

- [x] #237
- [ ] nonces
- [ ] block numbers

### PR DESCRIPTION
**Pull request by cptartur**
Closes #104 

## Introduced changes

-
-

## Checklist

- [x] Linked relevant issue
- [x] Updated relevant documentation
- [x] Added relevant tests
- [x] Performed self-review of the code

### CODE REVIEW

File: starknet-foundry/crates/forge/tests/integration/starknet.rs [hunk 1]
```diff
@@ -0,0 +1,432 @@
...
+    )
+    .unwrap();
+
+    assert_passed!(result);
+}
+
+// TODO Make nonce behavior consistent with Starknet
+#[ignore]
+#[test]
+fn nonce_increases_between_transactions() {
```
Comment (cptartur) [inline starknet-foundry/crates/forge/tests/integration/starknet.rs:211]: Right now, nonces are inconsistent with starknet

File: starknet-foundry/crates/forge/tests/integration/starknet.rs [hunk 2]
```diff
@@ -0,0 +1,432 @@
...
+    .unwrap();
+
+    assert_passed!(result);
+}
+
+// TODO Make nonce behavior consistent with Starknet
+#[ignore]
+#[allow(clippy::too_many_lines)]
+#[test]
+fn nonce_increases_between_deploys_and_declares() {
```
Comment (cptartur) [inline starknet-foundry/crates/forge/tests/integration/starknet.rs:304]: Same here

**Review by piotmag769 — COMMENTED** [review]
I know it is out of scope of this PR, but can we move `TestFileSummary` to different file? I think I requested it a few PRs ago 🥺

File: starknet-foundry/crates/forge/tests/integration/starknet.rs [hunk 3]
```diff
@@ -0,0 +1,432 @@
...
+                let block_timestamp = dispatcher.read_block_timestamp();
+                let sequencer_address = dispatcher.read_sequencer_address();
+
+                dispatcher.write_block();
+                let next_block_number = dispatcher.read_block_number();
+                let next_block_timestamp = dispatcher.read_block_timestamp();
+                let next_sequencer_address = dispatcher.read_sequencer_address();
+
+                assert(next_block_number >= block_number, 'block number decrese');
+                assert(next_block_timestamp >= block_timestamp, 'block timestamp decrese');
```
Comment (piotmag769) [inline starknet-foundry/crates/forge/tests/integration/starknet.rs:136]: decrese -> decreases

File: starknet-foundry/crates/forge/tests/integration/starknet.rs [hunk 4]
```diff
@@ -0,0 +1,432 @@
...
+        Some(&Utf8PathBuf::from_path_buf(corelib().to_path_buf()).unwrap()),
+        &test.contracts(corelib().path()).unwrap(),
+        &Utf8PathBuf::from_path_buf(predeployed_contracts().to_path_buf()).unwrap(),
+    )
+    .unwrap();
+
+    assert_passed!(result);
+}
+
+// TODO Make nonce behavior consistent with Starknet
```
Comment (piotmag769) [inline starknet-foundry/crates/forge/tests/integration/starknet.rs:208]: Can we link this issue here?
https://github.com/foundry-rs/starknet-foundry/issues/258

**Review by piotmag769 — APPROVED** [review]

**Comment by cptartur** [comment]
> I know it is out of scope of this PR, but can we move `TestFileSummary` to different file? I think I requested it a few PRs ago 🥺

I wouldn't do this in this PR, but I'm happy to create a follow up with refactors

File: starknet-foundry/crates/forge/tests/integration/starknet.rs (hunk 4 above)
Comment (cptartur) [reply starknet-foundry/crates/forge/tests/integration/starknet.rs:208]: Creating #292 and linking it. #258 is cheatable starknet related.

**Comment by cptartur** [comment]
#294 294

**Review by piotmag769 — APPROVED** [review]

**Review by Radinyn — APPROVED** [review]
In the future this should be tested by `Cheatnet`

**Comment by cptartur** [comment]
> In the future this should be tested by `Cheatnet`

Should we create an issue in cheatnet for migrating these tests ?

**Comment by Radinyn** [comment]
> > In the future this should be tested by `Cheatnet`
> 
> Should we create an issue in cheatnet for migrating these tests ?

I think there's #305 encapsulates it

**Pull request merged**